crates/polars-stream/src/nodes/io_sources/ipc.rs

```diff
@@ -0,0 +1,436 @@
+                        // left.
+                        None => is_batch_complete = true,
+                        Some(record_batch) => {
+                            let rb_num_rows = record_batch?.length()? as usize;
+                            batch.num_rows += rb_num_rows;
+
+                            // We need to ensure that we are not overflowing the IdxSize maximum
+                            // capacity.
+                            let rb_num_rows = IdxSize::try_from(rb_num_rows)
+                                .map_err(|_| ROW_COUNT_OVERFLOW_ERR)?;
+                            row_idx_offset = row_idx_offset
+                                .checked_add(rb_num_rows)
+                                .ok_or(ROW_COUNT_OVERFLOW_ERR)?;
+                        },
+                    }
+
+                    let current_block = reader.get_current_block();
+
+                    // Subdivide into batches for large files
+                    is_batch_complete |= batch.num_rows >= batch_size_limit;
+                    // Subdivide into batches if the file is sliced
+                    is_batch_complete |= batch.num_rows >= sliced_batch_size_limit;
+                    // Subdivide into batches for small files
+                    is_batch_complete |= current_block - batch.block_start >= batch_block_limit;
+
+                    // Batch blocks such that we send appropriately sized morsels. We guarantee a
+                    // lower bound here, but not an upperbound.
+                    if is_batch_complete {
+                        let batch_slice = slice_take(&mut slice, batch.num_rows);
+                        let block_range = batch.block_start..current_block;
+
+                        let message = BatchMessage {
+                            source_idx,
+                            memslice: memslice.clone(),
+                            metadata: metadata_arc.clone(),
+                            file_path: file_path.clone(),
+                            row_idx_offset: batch.row_idx_offset,
+                            slice: batch_slice,
+                            seq,
+                            block_range,
+                        };
+
+                        if distribute.send(message).await.is_err() {
+                            break;
+                        };
+
+                        batch = Batch {
+                            row_idx_offset,
+                            block_start: current_block,
+                            num_rows: 0,
+                        };
+                        seq += 1;
+                    }
+                }
+
+                (data_scratch, message_scratch) = reader.get_scratches();
+            }
+
+            drop(distribute); // Inform decoder tasks to stop
+            for decoder_task in decoder_tasks {
+                decoder_task.await?;
+            }
+            *is_finished = true; // Inform the graph that the port is done.
```

Comment:
Uh, you don't know yet if you're finished. Remember, if something requests an early stop through the `stop_requested` in the `SourceToken` you must be able to continue where you left off in the next phase.